### DISCUSSION TITLE
type: object without properties defined

### DISCUSSION THREAD

**Discussion by mjt2021** [hotmeteor/spectator#86] [Q&A]
It would seem Spectator forces properties to be defined. Is there any workaround? Tx!

**Comment by jarrodparkes** [comment]
@mjt2021 can you post an example? Spectator shouldn't force properties to be defined unless the spec says they are required. Spectator can also fail if you specify a property as non-nullable, but it is set to `null`

**Reply by mjt2021** [reply]
hey there! thanks for your response. sure, there is the conflicting piece of spec
      
      components:
        schemas:
          PaginationResponse:
            title: Pagination Response
            allOf:
              - $ref: './Entities.yaml#/components/schemas/Pagination'
          BaseResponse:
            title: Base placeholder for all responses
            type: object
            required:
            - response
            properties:
              response:
                type: object
          ResponseWithPagination:
            title: Base response w pagination
            allOf:
              - $ref: '#/components/schemas/BaseResponse'
              - type: object
              - required:
                - paging
                properties:
                  paging:
                    $ref: '#/components/schemas/PaginationResponse'
          BulkEntityOperationResponse:

it fails specifically with

BaseResponse

tx!

**Reply by jarrodparkes** [reply]
@mjt2021 would you be able to post the failure — the one produced by the assertion?

**Reply by mjt2021** [reply]
i was just getting a 500 when calling getJson

i was able to track down the problem here
        
        array:51 [
          0 => array:5 [
            "file" => "/var/www/html/vendor/hotmeteor/spectator/src/Exceptions/SchemaValidationException.php"
            "line" => 264
            "function" => "handleError"
            "class" => "Illuminate\Foundation\Bootstrap\HandleExceptions"
            "type" => "->"
          ]
          1 => array:5 [
            "file" => "/var/www/html/vendor/hotmeteor/spectator/src/Exceptions/SchemaValidationException.php"
            "line" => 266
            "function" => "formatSchema"
            "class" => "Spectator\Exceptions\SchemaValidationException"
            "type" => "::"
          ]
          2 => array:5 [
            "file" => "/var/www/html/vendor/hotmeteor/spectator/src/Exceptions/SchemaValidationException.php"
            "line" => 196
            "function" => "formatSchema"
            "class" => "Spectator\Exceptions\SchemaValidationException"
            "type" => "::"
          ]
          3 => array:5 [
            "file" => "/var/www/html/vendor/hotmeteor/spectator/src/Exceptions/SchemaValidationException.php"
            "line" => 196
            "function" => "formatSchema"
            "class" => "Spectator\Exceptions\SchemaValidationException"
            "type" => "::"
          ]
          4 => array:5 [
            "file" => "/var/www/html/vendor/hotmeteor/spectator/src/Exceptions/SchemaValidationException.php"
            "line" => 78
            "function" => "formatSchema"
            "class" => "Spectator\Exceptions\SchemaValidationException"
            "type" => "::"
          ]
        ............................
          46 => array:5 [
            "file" => "/var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php"
            "line" => 678
            "function" => "run"
            "class" => "PHPUnit\Framework\TestSuite"
            "type" => "->"
          ]
          47 => array:5 [
            "file" => "/var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php"
            "line" => 670
            "function" => "run"
            "class" => "PHPUnit\Framework\TestSuite"
            "type" => "->"
          ]
          48 => array:5 [
            "file" => "/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php"
            "line" => 143
            "function" => "run"
            "class" => "PHPUnit\TextUI\TestRunner"
            "type" => "->"
          ]
          49 => array:5 [
            "file" => "/var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php"
            "line" => 96
            "function" => "run"
            "class" => "PHPUnit\TextUI\Command"
            "type" => "->"
          ]
          50 => array:5 [
            "file" => "/var/www/html/vendor/phpunit/phpunit/phpunit"
            "line" => 92
            "function" => "main"
            "class" => "PHPUnit\TextUI\Command"
            "type" => "::"
          ]
        ]


This is the error i saw:
`Undefined index: properties`

Lines 237-272 of SchemaValidationException.php
        
        // handle each schema type
                    foreach ($types as $type) {
                        switch ($type) {
                            case 'object':
                                // does this object support additional properties?
                                $additional_properties = true;
                                if (isset($schema['additionalProperties'])) {
                                    if (is_bool($schema['additionalProperties'])) {
                                        $additional_properties = $schema['additionalProperties'];
                                    }
                                }
        
                                if ($additional_properties) {
                                    $type_modifier = '++'.$type_modifier;
                                }
        
                                // create entry for object schema
                                $display_string = self::schemaItemDisplayString(
                                    'object',
                                    $type_modifier,
                                    $key_current,
                                    $key_modifier
                                );
                                $schema_map[$location_current] = self::indentedDisplayString($display_string, $indent_level);
        
                                // create entries for all object properties
                                $indent_level = ++$indent_level;
                                
                                foreach ($schema['properties'] as $key => $next_schema) {   <<<<< no verification if properties are required
                                    if (isset($schema['required'])) {
                                        $schema_map = array_merge($schema_map, self::formatSchema($next_schema, $location_current, $key, $schema['required'], $indent_level));
                                    } else {
                                        $schema_map = array_merge($schema_map, self::formatSchema($next_schema, $location_current, $key, [], $indent_level));
                                    }
                                }
                                break;

**Reply by mjt2021** [reply]
sorry, thank you so much for your time! awesome lib :)

**Reply by jarrodparkes** [reply]
@mjt2021 

So in the `foreach ($schema['properties'] as $key => $next_schema)`, the error was occurring because `properties` was not defined? That is what I'm gathering from your response. Did you fix this by changing your spec or is this something we should be checking for in the library itself?

**Reply by mjt2021** [reply]
yes, exactly

right now i'm trying to clean up our spec, for the time being. 

i'm fairly new to the openapi concept so i am not sure about the implication of having an object without any defined properties, maybe that is a fault on our end

**Reply by jarrodparkes** [reply]
I'm also unsure whether...

1. `properties` should always be required from OAS perspective
2. that method should have even been called for the schema being analyzed
3. we just put a safety `isset(...)` check to avoid your issue

It sounds like all are valid options right now.
